### PR TITLE
feat(prefs): add iOS impl + non-effect interface

### PR DESCRIPTION
**Pull request by liambutler-lawrence**

### CODE REVIEW

File: packages/npm/prefs/src/index.tsx [hunk 1]
```diff
@@ -25,102 +25,107 @@ const prefsNativeModule: {
...
-export const S_CNDL_Int = S.Int.pipe(S.brand('CNDL_Int'))
+export const S_CNDL_UUID = Schema.UUID.pipe(Schema.brand('CNDL_UUID'))
+export const S_CNDL_Int = Schema.Int.pipe(Schema.brand('CNDL_Int'))
 
-export const PrefsError = S.Struct({
-  userInfo: S.OptionFromNullOr(S.Any),
-  nativeStackAndroid: S.OptionFromUndefinedOr(
-    S.Array(
-      S.Struct({
+export class PrefsError extends Schema.Class<PrefsError>('PrefsError')({
```
Comment (nikwotton) [inline packages/npm/prefs/src/index.tsx:31]: A fan of classes now?

**Review by nikwotton — APPROVED** [review]

File: packages/npm/prefs/src/index.tsx (hunk 1 above)
Comment (liambutler-lawrence) [reply packages/npm/prefs/src/index.tsx:31]: Haha no but since we're exposing the Promise-based interface, the traditional JS check for narrowing thrown errors is `error instanceof SomeErrorClass`, so figured it was worth the tradeoff.

File: packages/npm/prefs/src/index.tsx [hunk 2]
```diff
@@ -25,102 +25,107 @@ const prefsNativeModule: {
...
 
 // NOTE: The React Native bridge stringifies all objects anyway; we just do it manually so we can take advantage of Swift.Codable and kotlinx.serialization to simplify the native code
 // FIXME: Also validate that output is a string?
 export const getPref = (
   key: string
-): E.Effect<
-  O.Option<string>,
+): Effect.Effect<
+  Option.Option<string>,
   Discriminated<
```
Comment (gtokman) [inline packages/npm/prefs/src/index.tsx:73]: Option.Option? why the redundancy?

File: packages/npm/prefs/src/index.tsx [hunk 3]
```diff
@@ -25,102 +25,107 @@ const prefsNativeModule: {
...
-    E.map(O.fromNullable)
+    Effect.map(Option.fromNullable)
   )
 
 export const setPref = (
   key: string,
   value: string
-): E.Effect<void, typeof PrefsError.Type> =>
+): Effect.Effect<void, PrefsError> =>
   pipe(
```
Comment (gtokman) [inline packages/npm/prefs/src/index.tsx:112]: Same here, can you ditch the namespace & just import { Effect } from '...'

File: packages/npm/prefs/ios/Prefs.swift [hunk 4]
```diff
@@ -1,393 +1,27 @@
...
+        guard let stringValue = anyValue as? String? else {
+            return reject(CODE_NON_STRING_VALUE, MESSAGE_NON_STRING_VALUE, nil)
+        }
+        resolve(stringValue)
+    }
+    
+    @objc(setPref:toValue:withResolver:withRejecter:)
+    func setPref(key: String, toValue value: String, resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
+        UserDefaults.standard.set(value, forKey: key)
+        resolve(nil)
```
Comment (gtokman) [inline packages/npm/prefs/ios/Prefs.swift:19]: My preference would be to create our own default and inject the shared group if there is one otherwise we can use the default. 

We will need it if we develop widgets, notification extensions, or other targets in the future. 

```
if let userDefaults = UserDefaults(suiteName: "group.your.bundle.here") {
    userDefaults.set("test 1" as AnyObject, forKey: "key1")
    userDefaults.set("test 2" as AnyObject, forKey: "key2")
    userDefaults.synchronize()
}
```

File: packages/npm/prefs/src/index.tsx [hunk 5]
```diff
@@ -25,102 +25,107 @@ const prefsNativeModule: {
...
           error,
-          S.decodeUnknownEither(PrefsError),
+          Schema.decodeUnknownEither(PrefsError),
           Either.getRight,
-          O.flatMap(
+          Option.flatMap(
             oLiftRefinement(
               discriminateA('code', [
                 '@candlefinance.prefs.unexpected',
                 '@candlefinance.prefs.edit_commit_failed',
```
Comment (gtokman) [inline packages/npm/prefs/src/index.tsx:124]: #? what does edit commit failed mean?

File: packages/npm/prefs/ios/Prefs.swift [hunk 6]
```diff
@@ -1,393 +1,27 @@
...
-                }
-                resolve(stringifiedAccountBalanceList)
-            } catch let error as UnknownCurrentBalanceTypeError {
-                reject(CODE_UNKNOWN_CURRENT_BALANCE_TYPE, error.message, error)
-            } catch let error {
-                reject(CODE_UNKNOWN, MESSAGE_UNKNOWN, error)
-            }
-        }
+    
+    @objc(getPref:withResolver:withRejecter:)
```
Comment (gtokman) [inline packages/npm/prefs/ios/Prefs.swift:7]: #req `final`

File: packages/npm/prefs/README.md [hunk 7]
```diff
@@ -10,32 +10,29 @@ npm install @candlefinance/prefs
 
 ## Usage
 
-This package is optimized for usage with [Effect](https://effect.website), but can also be used in any React Native project.
+This package is designed for usage with [Effect](https://effect.website), but also exposes a Promise-based interface that can be used in any React Native project.
```
Comment (gtokman) [inline packages/npm/prefs/README.md:14]: 1. Can you update the installation instructions cmd to use `yarn` as that is more common in RN, or add both? 
2. To use this package, the end user must install Effect and the pollyfills. I would mention that and add that to the install cmd, anything you had to add to the example app will be req as a dependency.

File: packages/npm/financekit/example/ios/Podfile.lock [hunk 8]
```diff
@@ -1370,4 +1370,4 @@ SPEC CHECKSUMS:
 
 PODFILE CHECKSUM: 978a9588449ad9b8f7fc872a181d9140e7705193
 
-COCOAPODS: 1.15.2
+COCOAPODS: 1.14.3
```
Comment (gtokman) [inline packages/npm/financekit/example/ios/Podfile.lock:1373]: #? can you revert the changes to financekit

File: packages/npm/prefs/src/index.tsx [hunk 9]
```diff
@@ -25,102 +25,107 @@ const prefsNativeModule: {
...
   pipe(
-    E.tryPromise({
+    Effect.tryPromise({
       try: () => prefsNativeModule.getPref(key),
       catch: (error) =>
         pipe(
           error,
-          S.decodeUnknownEither(PrefsError),
+          Schema.decodeUnknownEither(PrefsError),
           Either.getRight,
```
Comment (gtokman) [inline packages/npm/prefs/src/index.tsx:88]: #? Looks like you're just wrapping a try/catch. Wouldn't it be easier to just use JS try/catch or better yet, return `prefsNativeModule.getPref(key)` and let the caller do this however they prefer?

File: packages/npm/prefs/README.md [hunk 10]
```diff
@@ -10,32 +10,29 @@ npm install @candlefinance/prefs
...
-To use `prefs` **with** Effect, check out the [example app](example/src/App.tsx/).
+To use `prefs` **with** Effect, import `@candlefinance/prefs`. Sample code is available in the [example app](example/src/App.tsx/).
 
-To use `prefs` **without** Effect, you can use the following bridging functions:
+To use `prefs` **without** Effect, import `@candlefinance/prefs/promises`. Functions throw a `PrefError` on unexpected issues.
 
 ```ts
 import { Effect, Option } from 'effect'
-import { getPref, setPref, deletePref } from '@candlefinance/prefs'
+import { getPref, setPref, deletePref } from '@candlefinance/prefs/promises'
```
Comment (gtokman) [inline packages/npm/prefs/README.md:22]: #? this seems to be a weird way to import something why req this path?

**Review by gtokman — CHANGES_REQUESTED** [review]

File: packages/npm/prefs/src/index.tsx (hunk 2 above)
Comment (nikwotton) [reply packages/npm/prefs/src/index.tsx:73]: Effect standard - `Option` is the file, `Option.Option` is the namespace. We shorten it to `O.Option` in our repo but we're not following the convention doing that

Comment (liambutler-lawrence) [reply packages/npm/prefs/src/index.tsx:73]: Almost haha, Option is the file/namespace, Option.Option is the actual type

File: packages/npm/prefs/src/index.tsx (hunk 5 above)
Comment (liambutler-lawrence) [reply packages/npm/prefs/src/index.tsx:124]: On the Android side you have to "commit" write operations, and they're failable.

File: packages/npm/prefs/ios/Prefs.swift (hunk 6 above)
Comment (liambutler-lawrence) [reply packages/npm/prefs/ios/Prefs.swift:7]: Say more please 😅

File: packages/npm/prefs/src/index.tsx (hunk 9 above)
Comment (liambutler-lawrence) [reply packages/npm/prefs/src/index.tsx:88]: No, because then we wouldn't be returning a PrefError. The Effect code lets us concisely handle potential errors and massage them  into a) fully typed error unions in the Effect interface, and b) still mostly typed error unions in the Promise-based interface (users just have to check `e instanceof PrefError`. 
That's as opposed to throwing random errors and making the user catch the RN thrown errors which aren't even defined as a class 😑

Comment (liambutler-lawrence) [reply packages/npm/prefs/src/index.tsx:88]: This is the fix for the mess you're dealing with in #oss right now where we're dealing with other library's errors and can't really help library consumers. When we properly handle and type errors, we can just respond with "cast to `instanceof PrefError` and read the code and message"

File: packages/npm/prefs/README.md (hunk 7 above)
Comment (liambutler-lawrence) [reply packages/npm/prefs/README.md:14]: Will add yarn command to all 3 packages, good catch.

Effect is a transitive dependency, doesn't have to be installed explicitly by the user. Will add note about polyfill, though I imagine most RN apps already have a TextEncoder polyfill.

File: packages/npm/financekit/example/ios/Podfile.lock (hunk 8 above)
Comment (liambutler-lawrence) [reply packages/npm/financekit/example/ios/Podfile.lock:1373]: This is an intentional fix, cocoapods is set to 1.14.3 in the Gemfile.lock. I think @nikwotton accidentally ran it without `bundle exec` in his last PR

File: packages/npm/prefs/ios/Prefs.mm [hunk 11]
```diff
@@ -2,17 +2,9 @@
...
-RCT_EXTERN_METHOD(transactionHistory:(NSString *)stringifiedParams withResolver:(RCTPromiseResolveBlock)resolve withRejecter:(RCTPromiseRejectBlock)reject)
-
-RCT_EXTERN_METHOD(accounts:(NSString *)stringifiedQuery withResolver:(RCTPromiseResolveBlock)resolve withRejecter:(RCTPromiseRejectBlock)reject)
-RCT_EXTERN_METHOD(accountHistory:(NSString *)stringifiedParams withResolver:(RCTPromiseResolveBlock)resolve withRejecter:(RCTPromiseRejectBlock)reject)
-
-RCT_EXTERN_METHOD(accountBalances:(NSString *)stringifiedQuery withResolver:(RCTPromiseResolveBlock)resolve withRejecter:(RCTPromiseRejectBlock)reject)
-RCT_EXTERN_METHOD(accountBalanceHistory:(NSString *)stringifiedParams withResolver:(RCTPromiseResolveBlock)resolve withRejecter:(RCTPromiseRejectBlock)reject)
+RCT_EXTERN_METHOD(getPref:(NSString *)key withResolver:(RCTPromiseResolveBlock)resolve withRejecter:(RCTPromiseRejectBlock)reject)
+RCT_EXTERN_METHOD(setPref:(NSString *)key toValue:(NSString *)value withResolver:(RCTPromiseResolveBlock)resolve withRejecter:(RCTPromiseRejectBlock)reject)
+RCT_EXTERN_METHOD(deletePref:(NSString *)key withResolver:(RCTPromiseResolveBlock)resolve withRejecter:(RCTPromiseRejectBlock)reject)
```
[the remark this answers is not in the document]
Comment (liambutler-lawrence) [reply packages/npm/prefs/ios/Prefs.mm:7]: My understanding is using synchronous methods over the (old) RN bridge is bad

https://reactnative.dev/docs/0.72/native-modules-ios

> At the moment, we do not recommend using synchronous methods, since calling methods synchronously can have strong performance penalties and introduce threading-related bugs to your native modules.

File: packages/npm/prefs/ios/Prefs.swift (hunk 4 above)
Comment (liambutler-lawrence) [reply packages/npm/prefs/ios/Prefs.swift:19]: Happy to do this, Android has a similar parameter. But how do you imagine injecting in the suite/file name? Looked at RN docs but didn't see any official way to pass config/initialization parameters to native modules.

Comment (gtokman) [reply packages/npm/prefs/ios/Prefs.swift:19]: Inject it into the function as an optional parameter.

File: packages/npm/prefs/ios/Prefs.swift (hunk 6 above)
Comment (gtokman) [reply packages/npm/prefs/ios/Prefs.swift:7]: `final class Pref`

File: packages/npm/prefs/ios/Prefs.mm (hunk 11 above)
[the remark this answers is not in the document]
Comment (gtokman) [reply packages/npm/prefs/ios/Prefs.mm:7]: that was docs for 0.72 and in practice it's not bad.

[the remark this answers is not in the document]
Comment (liambutler-lawrence) [reply packages/npm/prefs/ios/Prefs.mm:7]: Sorry, wrong link: https://reactnative.dev/docs/native-modules-ios The latest docs still have the exact same copy.

It's not bad based on what?

File: packages/npm/prefs/ios/Prefs.swift (hunk 6 above)
Comment (liambutler-lawrence) [reply packages/npm/prefs/ios/Prefs.swift:7]: Cool, updated

File: packages/npm/prefs/ios/Prefs.swift (hunk 4 above)
Comment (liambutler-lawrence) [reply packages/npm/prefs/ios/Prefs.swift:19]: Ah, that seems not ideal... we'd have to dynamically create UserDefaults and SharedPreferences instances on each method call 🤔 would rather punt this since it's not something we need today

File: packages/npm/prefs/ios/Prefs.mm (hunk 11 above)
[the remark this answers is not in the document]
Comment (candle-admin) [reply packages/npm/prefs/ios/Prefs.mm:7]: In practice I haven't seen any issue from using the sync functions. I think if you have them do compute intensive tasks they are cause a performance hit on the JS thread.

File: packages/npm/prefs/README.md [hunk 12]
```diff
@@ -30,7 +54,8 @@ await deletePref('themeColor')
 
 ### Important Notes
 
-- SharedPreferences on Android does not expose a default app-wide instance (like UserDefaults on iOS). To a) access prefs saved using this library from native Android code, or b) write prefs from native Android code and then access them using this library, initialize SharedPreferences in your native Android code with the file name `com.candlefinance.prefs`.
+- Effect requires that `TextEncoder`/`TextDecoder` be defined at runtime. If you don't already have a polyfill, we recommend installing [fast-text-encoding](https://github.com/samthor/fast-text-encoding) using NPM or Yarn and adding `import 'fast-text-encoding'` at the top of your app's `index.js` file.
+- SharedPreferences on Android does not expose a default app-wide instance (like UserDefaults on iOS). To a) access prefs saved using this library from native Android code, or b) write prefs from native Android code and then access them using this library, [initialize SharedPreferences](https://developer.android.com/training/data-storage/shared-preferences) in your native Android code with the file name `com.candlefinance.prefs`.
```
Comment (candle-admin) [inline packages/npm/prefs/README.md:58]: can you update the install to `yarn add @candlefinance/prefs fast-text-encoding`

File: packages/npm/prefs/ios/Prefs.swift (hunk 4 above)
Comment (gtokman) [reply packages/npm/prefs/ios/Prefs.swift:19]: It's just initializing a class with a shared group. I don't think that's bad. But cool w/ punting.

File: packages/npm/prefs/ios/Prefs.mm (hunk 11 above)
[the remark this answers is not in the document]
Comment (gtokman) [reply packages/npm/prefs/ios/Prefs.mm:7]: In practice I haven't seen any issue from using the sync functions. I think if you have them do compute intensive tasks they are cause a performance hit on the JS thread.

File: packages/npm/prefs/README.md [hunk 13]
```diff
@@ -1,23 +1,47 @@
...
 ```sh
 npm install @candlefinance/prefs
 ```
 
+To install `prefs` with **yarn**:
+
+```sh
+yarn add @candlefinance/prefs
+```
+
```
Comment (gtokman) [inline packages/npm/prefs/README.md:31]: yarn add @candlefinance/prefs fast-text-encoding

**Review by gtokman — APPROVED** [review]

File: packages/npm/prefs/README.md (hunk 13 above)
Comment (liambutler-lawrence) [reply packages/npm/prefs/README.md:31]: I added a note about that below

File: packages/npm/prefs/src/index.tsx (hunk 5 above)
Comment (liambutler-lawrence) [reply packages/npm/prefs/src/index.tsx]: Updated this code to be non-platform-specific

File: packages/npm/prefs/ios/Prefs.mm (hunk 11 above)
[the remark this answers is not in the document]
Comment (liambutler-lawrence) [reply packages/npm/prefs/ios/Prefs.mm:7]: Ah, as in the sync methods run directly on the RN "JS" thread as opposed to the async functions which are run on a separate queue? I guess my question is, is there any downside to using the async versions, even if the performance difference is negligible?

Either way, going to punt for now.
